llvm/lib/Passes/PassBuilderPipelines.cpp

```diff
@@ -1629,6 +1630,23 @@ PassBuilder::buildModuleOptimizationPipeline(OptimizationLevel Level,
   if (!LTOPreLink)
     MPM.addPass(RelLookupTableConverterPass());
 
+  if (PTO.WholeProgramDevirt && LTOPhase == ThinOrFullLTOPhase::None) {
```

Comment:
Does this need to be guarded by whether whole program vtables was enabled? We don't for LTO. What happens if it is simply always invoked for non-LTO?

Reply:
At least WPD and LTT should not do anything if there is no type metadata (the inliner invocation is a separate story, more comments about that below).